source/libraries/ContractDeployer.ts

```diff
@@ -0,0 +1,231 @@
+
+        //
+        // if [ -z $SAI_ADM ]
+        // then
+        //     SAI_ADM=$(dapp create DSRoles)
+        //     seth send $SAI_ADM 'setRootUser(address,bool)' $ETH_FROM true
+        // fi
+        const saiAdmContract = new DSRoles(this.connector, this.accountManager, await this.simpleDeploy("DSRoles"), this.connector.gasPrice)
+
+        // seth send $DAI_FAB 'makeTokens()
+        console.log("DaiFab.makeTokens()");
+        await daiFabContract.makeTokens();
+
+        // seth send $DAI_FAB 'makeVoxTub(address,address,address,address,address)' $SAI_GEM $SAI_GOV $SAI_PIP $SAI_PEP $SAI_PIT
+        console.log("DaiFab.makeVoxTub()");
+        await daiFabContract.makeVoxTub(
+            saiGemContract.address,
+            saiGovContract.address,
+            saiPipContract.address,
+            saiPepContract.address,
+            saiPitAddress,
+        );
+
+        // seth send $DAI_FAB 'makeTapTop()'
+        console.log("DaiFab.makeTapTop()");
+        await daiFabContract.makeTapTop();
+
+        // seth send $DAI_FAB 'configParams()'
+        console.log("DaiFab.configParams()");
+        await daiFabContract.configParams();
+
+        console.log("DaiFab.verifyParams()");
+        // seth send $DAI_FAB 'verifyParams()'
+        await daiFabContract.verifyParams();
+
+        // seth send $DAI_FAB 'configAuth(address)' $SAI_ADM
+        console.log("DaiFab.configAuth()");
+        await daiFabContract.configAuth(saiAdmContract.address);
+
+        const deployedContractAddresses = {
+            gem: saiGemContract.address,
+            gov: saiGovContract.address,
+            pip: saiPipContract.address,
+            pep: saiPepContract.address,
+            pit: saiPitAddress,
+            adm: saiGemContract.address,
+
+            sai: await daiFabContract.sai_(),
+            sin: await daiFabContract.sin_(),
+            skr: await daiFabContract.skr_(),
+            dad: await daiFabContract.dad_(),
+            mom: await daiFabContract.mom_(),
+            vox: await daiFabContract.vox_(),
+            tub: await daiFabContract.tub_(),
+            tap: await daiFabContract.tap_(),
+            top: await daiFabContract.top_(),
+        };
+        await this.generateAddressMappingFile(deployedContractAddresses);
+    }
+
+    private async simpleDeploy(contractName: string, constructorArgs?: Array<any>): Promise<string> {
+        return await this.construct(this.contracts.get(contractName), constructorArgs || [], `Uploading ${contractName}`);
+    }
+
+    private static getEncodedConstructData(abi: Abi, bytecode: Buffer, constructorArgs: Array<string>): Buffer {
+        if (constructorArgs.length === 0) {
+            return bytecode;
+        }
+        // TODO: submit a TypeScript bug that it can't deduce the type is AbiFunction|undefined here
+        const constructorSignature = <AbiFunction | undefined>abi.find(signature => signature.type === 'constructor');
```

Comment:
Recommend verifying that this is still a bug (unless this comment is from you, though it sounds like one of mine).

Reply:
It's not my comment, i'll check on this and the gas one

Reply:
It looks like it isn't a bug, type infers you could be returning an AbiEvent as well.

Reply:
Ah, I think the issue is that `signature.type === 'constructor' is only valid when `signature` is of type `AbiFunction`.  You can see in the type definitions that you cannot be `AbiEvent` and have `type === 'constructor'`:
```typescript
export interface AbiFunction {
	name: string,
	type: 'function' | 'constructor' | 'fallback',
	stateMutability: 'pure' | 'view' | 'payable' | 'nonpayable',
	constant: boolean,
	payable: boolean,
	inputs: Array<AbiParameter>,
	outputs: Array<AbiParameter>,
}

export interface AbiEvent {
	name: string,
	type: 'event',
	inputs: Array<AbiEventParameter>,
	anonymous: boolean,
}
```

Reply:
I'm curious though, what if you changed it to this:
```typescript
function matcher(signature: AbiFunction | AbiEvent): signature is AbiFunction {
    return signature = signature.type === 'constructor'
}
const constructorSignature = abi.find(matcher)
```

Reply:
it's mixing booleans and AbiFunction. And either way you're coercing, it's just if you do it one level down

Reply:
So it does seem like a typescript bug, given that it should be able to deduce it would only return true for AbiFunction